python/ray/data/tests/test_dataset.py

```diff
@@ -186,6 +188,355 @@ def test_tensors(ray_start_regular_shared):
+        return 2 * (arr + 1) / 3
+
+    def apply_comparison_ops(arr):
+        return arr % 2 == 0
+
+    def apply_logical_ops(arr):
+        return arr & (3 * arr) | (5 * arr)
+
+    # Op tests, using NumPy as the groundtruth.
+    np.testing.assert_equal(
+        apply_arithmetic_ops(arr), apply_arithmetic_ops(df["two"]))
+
+    np.testing.assert_equal(
+        apply_comparison_ops(arr), apply_comparison_ops(df["two"]))
+
+    np.testing.assert_equal(
+        apply_logical_ops(arr), apply_logical_ops(df["two"]))
+
+
+def test_tensor_array_reductions(ray_start_regular_shared):
+    outer_dim = 3
+    inner_shape = (2, 2, 2)
+    shape = (outer_dim, ) + inner_shape
+    num_items = np.prod(np.array(shape))
+    arr = np.arange(num_items).reshape(shape)
+
+    df = pd.DataFrame({"one": list(range(outer_dim)), "two": TensorArray(arr)})
+
+    # Reduction tests, using NumPy as the groundtruth.
+    for name, reducer in TensorArray.SUPPORTED_REDUCERS.items():
+        np_kwargs = {}
+        if name in ("std", "var"):
+            # Pandas uses a ddof default of 1 while NumPy uses 0.
+            # Give NumPy a ddof kwarg of 1 in order to ensure equivalent
+            # standard deviation calculations.
+            np_kwargs["ddof"] = 1
+        np.testing.assert_equal(df["two"].agg(name),
+                                reducer(arr, axis=0, **np_kwargs))
+
+
+def test_tensors_in_tables_from_pandas(ray_start_regular_shared):
+    outer_dim = 3
+    inner_shape = (2, 2, 2)
+    shape = (outer_dim, ) + inner_shape
+    num_items = np.prod(np.array(shape))
+    arr = np.arange(num_items).reshape(shape)
+    df = pd.DataFrame({"one": list(range(outer_dim)), "two": list(arr)})
+    # Cast column to tensor extension dtype.
+    df["two"] = df["two"].astype(TensorDtype())
+    ds = ray.data.from_pandas([ray.put(df)])
+    values = [[s["one"], s["two"]] for s in ds.take()]
+    expected = list(zip(list(range(outer_dim)), arr))
+    for v, e in zip(sorted(values), expected):
+        np.testing.assert_equal(v, e)
+
+
+def test_tensors_in_tables_pandas_roundtrip(ray_start_regular_shared):
+    outer_dim = 3
+    inner_shape = (2, 2, 2)
+    shape = (outer_dim, ) + inner_shape
+    num_items = np.prod(np.array(shape))
+    arr = np.arange(num_items).reshape(shape)
+    df = pd.DataFrame({"one": list(range(outer_dim)), "two": TensorArray(arr)})
+    ds = ray.data.from_pandas([ray.put(df)])
+    ds_df = ray.get(ds.to_pandas())[0]
+    assert ds_df.equals(df)
+
+
+def test_tensors_in_tables_parquet_roundtrip(ray_start_regular_shared,
+                                             tmp_path):
+    outer_dim = 3
+    inner_shape = (2, 2, 2)
+    shape = (outer_dim, ) + inner_shape
+    num_items = np.prod(np.array(shape))
+    arr = np.arange(num_items).reshape(shape)
+    df = pd.DataFrame({"one": list(range(outer_dim)), "two": TensorArray(arr)})
+    ds = ray.data.from_pandas([ray.put(df)])
+    ds.write_parquet(str(tmp_path))
```

Comment:
Interesting, how does the parquet roundtrip work exactly (do you have links on how the parquet metadata is stored to allow this?)

Reply:
The storage arrays (data and offset buffers) are stored in Parquet under their storage type (tensor element dtype and `int32`, respectively), while the extension type metadata is saved within the serialized Arrow schema. So, when Arrow is reading the storage arrays, it knows from the Arrow schema to interpret those storage arrays as our extension type.